### PR TITLE
Dockerized server + Added Envoy submodule

### PR DESCRIPTION
**Pull request by Hedhav**
Dockerized server with services- golang server, mysql, phpmyadmin, dalal envoy proxy

 Added Envoy submodule

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -37,14 +37,14 @@ jobs:
           command: |
             sudo apt-get update
             sudo apt install -y default-mysql-client
-            mysql -h 127.0.0.1 -u root -e 'CREATE DATABASE dalalstreet_test;'
+            mysql -h 127.0.0.1 -u root -e 'CREATE DATABASE dalalstreet_docker;'
```
Comment (siva2204) [inline .circleci/config.yml:40]: change this back to `dalalstreet_test`

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,55 @@
+version: '2'
```
Comment (siva2204) [inline docker-compose.yml:1]: change this to version '3'

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+services:
+    
+    dalalserver:
+        container_name: dalalstreet_server_2
+        build: .
+        volumes: 
+            # - $GOPATH:/go
+            - ./:/go/src/github.com/delta/dalal-street-server
+        ports:
+            - "8000:8000"
```
Comment (siva2204) [inline docker-compose.yml:11]: dont hardcode ports in compose, use `.env`

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            MYSQL_DATABASE: ${DB_NAME}
+            MYSQL_ROOT_PASSWORD: ${DB_PASS}
+    
+    phpmyadmin:
+        container_name: dalalstreet_pma_2
+        image: phpmyadmin/phpmyadmin
+        env_file:
+            - .env
+        ports:
+            - "9040:80"
```
Comment (siva2204) [inline docker-compose.yml:40]: dont hardcode ports value

File: docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+    dalalenvoy:
+        container_name: dalal-envoy-2
+        build: 
+            context: ./dalal-envoy-proxy/
+            dockerfile: dockerfiles/Dockerfile.dev
+        ports:
+            - "3000:3000"
+        extra_hosts:
+            - host.docker.internal:host-gateway
```
Comment (siva2204) [inline docker-compose.yml:55]: dont think we need this, as all the services are in same network

File: docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+            PMA_HOST: db
+
+
+    dalalenvoy:
+        container_name: dalal-envoy-2
+        build: 
+            context: ./dalal-envoy-proxy/
+            dockerfile: dockerfiles/Dockerfile.dev
+        ports:
+            - "3000:3000"
```
Comment (siva2204) [inline docker-compose.yml:53]: dont hardcode ports

File: docker-entry.sh [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+echo "########## Building proto files ###########"
+export PATH=$PATH:/root/protobuf/bin
+protoc --version
+bash build_proto.sh
+
+
+
+
+# Get the database password
+dbPass=$(egrep "Docker|DbPassword" config.json \
```
Comment (siva2204) [inline docker-entry.sh:50]: we can use .env for this, by default compose will import `.env`

File: docker_server_setup.sh [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+
+
+echo "######## Fetching Go dependencies ##########"
+go install -tags 'mysql' github.com/golang-migrate/migrate/v4/cmd/migrate@latest
+# go get -tags 'mysql' -u github.com/golang-migrate/migrate/v4/cmd/migrate/
+go install google.golang.org/protobuf/cmd/protoc-gen-go@latest
+go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@v1.1
+
+echo "######## Go dependencies Fetched (Partial) ##########"
```
Comment (siva2204) [inline docker_server_setup.sh:23]: add new line here

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    netcat \
+    git
+
+WORKDIR  /go/src/github.com/delta/dalal-street-server 
+
+ADD ./docker_server_setup.sh ./docker_server_setup.sh
+RUN "./docker_server_setup.sh"
+# The above setup is run seperately earlier on, so that cache can be used 
+# when rebuilding the image, in case of any change
+
```
Comment (siva2204) [inline Dockerfile:18]: `COPY` go.mod go.sum ./

`RUN` go mod download

use this to install dependency

check [here](https://docs.docker.com/language/golang/build-images/) for more info

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+RUN apt-get update && \
+    apt-get install -y apt-utils \
+    zip \
+    unzip \
+    vim \
+    curl \
+    netcat \
+    git
+
+WORKDIR  /go/src/github.com/delta/dalal-street-server
```
Comment (siva2204) [inline Dockerfile:12]: as we are using go modules, we dont have to $GOPATH, but anything is fine

File: docker-compose.yml [hunk 11]
```diff
@@ -0,0 +1,55 @@
+version: '2'
+services:
+    
+    dalalserver:
+        container_name: dalalstreet_server_2
+        build: .
+        volumes: 
+            # - $GOPATH:/go
+            - ./:/go/src/github.com/delta/dalal-street-server
```
Comment (siva2204) [inline docker-compose.yml:9]: dont think we need to add volumes for this

File: Dockerfile [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+    zip \
+    unzip \
+    vim \
+    curl \
+    netcat \
+    git
+
+WORKDIR  /go/src/github.com/delta/dalal-street-server 
+
+ADD ./docker_server_setup.sh ./docker_server_setup.sh
```
Comment (siva2204) [inline Dockerfile:14]: can use `COPY` ?

File: docker-compose.yml [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+            DALAL_ENV: Docker
+        restart: always
+
+    db:
+        container_name: dalalstreet_db_2
+        image: mysql:5.7
+        volumes:
+            - ./docker/mysql:/var/lib/mysql
+        env_file:
+            - .env
```
Comment (siva2204) [inline docker-compose.yml:27]: where is sample env file ?

**Review by siva2204 — CHANGES_REQUESTED** [review]
- can rename `docker_server_setup.sh` to `server-setup.sh` ?
- container names can be `dalalstreet-x`. ex: dalalstreet-db

**Comment by siva2204** [comment]
- revert all the changes in go.sum and go.mod(`run go mod tidy` may be)

File: .env.example [hunk 14]
```diff
@@ -0,0 +1,6 @@
+DB_NAME=dalalstreet_docker
+DB_PASS=
+SERVER_PORT=8000
+MYSQL_PORT=3306
+PMA_PORT=9040
+ENVOY_PORT=3000
```
Comment (siva2204) [inline .env.example:6]: Add new line

File: docker-compose.yml [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+
+    db:
+        container_name: dalalstreet-db
+        image: mysql:5.7
+        volumes:
+            - ./docker/mysql:/var/lib/mysql
+        env_file:
+            - .env
+        ports:
+            - "${MYSQL_PORT}:3306"
```
Comment (siva2204) [inline docker-compose.yml:30]: we dont have to expose the mysql server port ?

Comment (siva2204) [reply docker-compose.yml:30]: remove MYSQL_PORT in .env

File: docker-entry.sh [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+    if [ $isopen -eq 0 ]; then
+        echo "#### No DB :(  #########"
+        break
+    fi
+    maxtry=maxtry-1
+    sleep 1
+done
+
+echo "######### Running migrations ##########"
+migrate -url "mysql://root:${MYSQL_ROOT_PASSWORD}@tcp(db:${MYSQL_PORT})/dalalstreet_docker" -path ./migrations up
```
Comment (siva2204) [inline docker-entry.sh:40]: migrate -path ./migrations -database "mysql://root:${MYSQL_ROOT_PASSWORD}@tcp(db)/dalalstreet_docker"  up

this is thy correct command

File: docker-compose.yml [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+        environment:
+            MYSQL_DATABASE: ${DB_NAME}
+            DALAL_ENV: Docker
+            MYSQL_ROOT_PASSWORD: ${DB_PASS}
+            MYSQL_PORT: ${MYSQL_PORT}
+        restart: always
+
+    db:
+        container_name: dalalstreet-db
+        image: mysql:5.7
```
Comment (siva2204) [inline docker-compose.yml:24]: try mysql:8 ?

File: docker-compose.yml [hunk 18]
```diff
@@ -0,0 +1,56 @@
+version: '3'
+services:
+    
+    dalalserver:
+        container_name: dalalstreet-server
+        build: .
+        volumes: 
+            - ./dalalstreet_docker.log:/dalal-street-server/dalalstreet_docker.log
```
Comment (anirudhvs) [inline docker-compose.yml:8]: Check the logs directory once. Refer (this)[https://github.com/delta/dalal-street-server/blob/d7b5145a593afb997f7599e73656843aac1cca54/docker-compose.yml#L18]

File: docker-compose.yml [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+        env_file:
+            - .env
+        environment:
+            MYSQL_DATABASE: ${DB_NAME}
+            DALAL_ENV: Docker
+            MYSQL_ROOT_PASSWORD: ${DB_PASS}
+            MYSQL_PORT: ${MYSQL_PORT}
+        restart: always
+
+    db:
```
Comment (anirudhvs) [inline docker-compose.yml:22]: set `restart: always`

Also write scripts to create and restore mysql dumps. Refer to prod2 branch.

File: docker-compose.yml [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+            - .env
+        ports:
+            - "3306:3306"
+        expose:
+            - "3306" 
+        environment:
+            MYSQL_DATABASE: ${DB_NAME}
+            MYSQL_ROOT_PASSWORD: ${DB_PASS}
+    
+    phpmyadmin:
```
Comment (anirudhvs) [inline docker-compose.yml:37]: Restart here too

File: docker-compose.yml [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+        container_name: dalalstreet-pma
+        image: phpmyadmin/phpmyadmin
+        env_file:
+            - .env
+        ports:
+            - "${PMA_PORT}:80"
+        depends_on:
+            - db
+        # environment:
+        #     PMA_HOST: db
```
Comment (anirudhvs) [inline docker-compose.yml:47]: Add this. Might be necessay in some cases. Add this in `.env` (preferred).

File: docker-entry.sh [hunk 22]
```diff
@@ -0,0 +1,34 @@
+#!/bin/sh
+
+set -x
```
Comment (anirudhvs) [inline docker-entry.sh:3]: Refer to the other comment about `set -x`

File: docker-entry.sh [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+
+
+# go get ./...
+# echo "######## Go dependencies Fetched (Full) ##########"
```
Comment (anirudhvs) [inline docker-entry.sh:7]: Remove commented code if not necessary

File: server-setup.sh [hunk 24]
```diff
@@ -0,0 +1,22 @@
...
+PB_REL="https://github.com/protocolbuffers/protobuf/releases"
+curl -LO $PB_REL/download/v3.15.8/protoc-3.15.8-linux-x86_64.zip
+
+
+
+echo "######## Unzipping protoc compiler ##########"
+unzip protoc-3.15.8-linux-x86_64.zip -d /root/protobuf
+
+echo "######## Adding to path ##########"
+export PATH=$PATH:/root/protobuf/bin
```
Comment (anirudhvs) [inline server-setup.sh:12]: Set in `Dockerfile`.

File: Dockerfile [hunk 25]
```diff
@@ -0,0 +1,27 @@
+FROM golang:1.16.9
+
+RUN apt-get update && \
+    apt-get install -y apt-utils \
+    zip \
+    unzip \
+    vim \
```
Comment (anirudhvs) [inline Dockerfile:7]: Vim is not necessary in production right?

File: docker-compose.yml [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+        container_name: dalalstreet-db
+        image: mysql:5.7
+        volumes:
+            - ./docker/mysql:/var/lib/mysql
+        env_file:
+            - .env
+        ports:
+            - "3306:3306"
+        expose:
+            - "3306"
```
Comment (anirudhvs) [inline docker-compose.yml:32]: Expose not necessary.

File: docker-compose.yml [hunk 27]
```diff
@@ -0,0 +1,56 @@
+version: '3'
+services:
+    
+    dalalserver:
+        container_name: dalalstreet-server
```
Comment (anirudhvs) [inline docker-compose.yml:5]: Add `image: dalalstreet_server` so it gets tagged properly after building

File: docker-compose.yml [hunk 28]
```diff
@@ -0,0 +1,56 @@
...
+        depends_on:
+            - db
+        env_file:
+            - .env
+        environment:
+            MYSQL_DATABASE: ${DB_NAME}
+            DALAL_ENV: Docker
+            MYSQL_ROOT_PASSWORD: ${DB_PASS}
+            MYSQL_PORT: ${MYSQL_PORT}
+        restart: always
```
Comment (anirudhvs) [inline docker-compose.yml:20]: put this in top after image and container name right? So everything looks consistent

**Review by anirudhvs — CHANGES_REQUESTED** [review]
Remove unnecessary comments and spaces.

File: docker-compose.yml (hunk 21 above)
Comment (Hedhav) [reply docker-compose.yml:47]: Uncommenting this part causes PMA to throw a "mysqli::real_connect(): (HY000/2002): Connection refused" error. I'm not sure how to resolve it. It was uncommented in my previous commits, and worked fine. Now, even if i hard reset to those commits, it isn't working anymore. Any ideas on what the issue is?

**Comment by siva2204** [comment]
- remove `MYSQL_PORT` in example.env and docker-compose.yaml file
- no need to initialise submodules in docker container
- remove unwanted pkg (like vim and git)
- change log volumes mount conf to something like this ` - ./docker/logs/:/dalal-street-server/logs`
- create a new directory and move all the script files into that ?

File: scripts/create_mysql_dump.sh [hunk 29]
```diff
@@ -0,0 +1,14 @@
...
+    mkdir -p dumps
+    time=$(date +%d-%m-%Y-%H-%M-%S)
+
+    if [ -f .env ]
+    then
+        export $(cat .env | xargs)
+    fi
+    docker exec dalalstreet-db /usr/bin/mysqldump -u root -p${DB_PASS} dalalstreet_docker > ./dumps/dump-$time.sql
+    
+fi
```
Comment (siva2204) [inline scripts/create_mysql_dump.sh:14]: add new line here

**Comment by siva2204** [comment]
add `.dockerignore` file

File: .dockerignore [hunk 30]
```diff
@@ -0,0 +1,16 @@
...
+.gitignore
+.gitmodules 
+
+# unnecessary files
+.env
+.env.example
+config.json.example
+README.md
+docker-compose.yml
+dockerfile
```
Comment (anirudhvs) [inline .dockerignore:16]: `Dockerfile`

File: docker-compose.yml [hunk 31]
```diff
@@ -0,0 +1,56 @@
...
+        ports:
+            - "${SERVER_PORT}:8000"
+        depends_on:
+            - db
+        env_file:
+            - .env
+        environment:
+            MYSQL_DATABASE: ${DB_NAME}
+            DALAL_ENV: Docker
+            MYSQL_ROOT_PASSWORD: ${DB_PASS}
```
Comment (anirudhvs) [inline docker-compose.yml:20]: Could you put everything in `.env` file itself? `DB_NAME` and `DB_PASS` are not being used anywhere. Directly put `MSQL_ROOT_DATABASE`, `MSQL_ROOT_PASSWORD` and `DALAL_ENV` in `.env` file and remove the environment section in compose file.

**Review by anirudhvs — CHANGES_REQUESTED** [review]

**Review by siva2204 — APPROVED** [review]

**Review by anirudhvs — APPROVED** [review]

**Pull request merged**